### ISSUE TITLE
1.12.0-rc3: Can't curl to service vip from inside container

### ISSUE THREAD

**Issue by jfbibeau** [docker/docker#24377]
**Steps to reproduce the issue:**
1. Init swarm (node a)
2. Join 2 nodes (node b, node c)
3. docker service create --name tomcat --publish 8180:8180 private-repo/tomcat
4. Exec bash into the container
5. curl -v tomcat:8180
-> Tries IP 10.255.0.6 (which is the VIP)  -> fails, No route to host
6. curl -v tasks.tomcat:8180
-> Tries IP 10.255.0.7 (which is the IP of the container) -> works.

**Describe the results you received:**
Can curl to the container IP from itself, but can't curl to the VIP
Additionally, if I start a new service on the same default network, and try to reach the container IP (tasks.tomcat) or the service VIP, neither work.

**Describe the results you expected:**
Should be able to curl to the VIP resolved by DNS...

**Additional information you deem important (e.g. issue happens only occasionally):**
100% reproducible. What am I missing?

**Comment by jfbibeau** [comment]
Output of network inspect for the default swarm network, and the service inspect:
```
[
    {
        "Name": "ingress",
        "Id": "aak1haq3dk71xyzj07kej1kni",
        "Scope": "swarm",
        "Driver": "overlay",
        "EnableIPv6": false,
        "IPAM": {
            "Driver": "default",
            "Options": null,
            "Config": [
                {
                    "Subnet": "10.255.0.0/16",
                    "Gateway": "10.255.0.1"
                }
            ]
        },
        "Internal": false,
        "Containers": {
            "aa8253f68fe3ff104c4ae542aa139f7f00a65ac36e4dff14dad0ec2bfef3fa77": {
                "Name": "tomcat.1.dcnryuws0pqqxg38oqpohwry1",
                "EndpointID": "c125187dae0ae4d1841b3abe82532ef52e1d4e4cc122f48923c9f086155be804",
                "MacAddress": "02:42:0a:ff:00:07",
                "IPv4Address": "10.255.0.7/16",
                "IPv6Address": ""
            },
            "ingress-sbox": {
                "Name": "ingress-endpoint",
                "EndpointID": "7bc725ee2c9ea91dd634a1efc7e47714dcdb8b2fa2c930c75623ab9c77c11474",
                "MacAddress": "02:42:0a:ff:00:03",
                "IPv4Address": "10.255.0.3/16",
                "IPv6Address": ""
            }
        },
        "Options": {
            "com.docker.network.driver.overlay.vxlanid_list": "256"
        },
        "Labels": {}
    }
]


[
    {
        "ID": "838joi5g94z3rlo1fk8wtbf3e",
        "Version": {
            "Index": 23
        },
        "CreatedAt": "2016-07-06T15:58:43.809500028Z",
        "UpdatedAt": "2016-07-06T15:58:43.815603001Z",
        "Spec": {
            "Name": "tomcat",
            "TaskTemplate": {
                "ContainerSpec": {
                    "Image": "private-repo/tomcat"
                },
                "Resources": {
                    "Limits": {},
                    "Reservations": {}
                },
                "RestartPolicy": {
                    "Condition": "any",
                    "MaxAttempts": 0
                },
                "Placement": {}
            },
            "Mode": {
                "Replicated": {
                    "Replicas": 1
                }
            },
            "UpdateConfig": {},
            "EndpointSpec": {
                "Mode": "vip",
                "Ports": [
                    {
                        "Protocol": "tcp",
                        "TargetPort": 8180,
                        "PublishedPort": 8180
                    }
                ]
            }
        },
        "Endpoint": {
            "Spec": {
                "Mode": "vip",
                "Ports": [
                    {
                        "Protocol": "tcp",
                        "TargetPort": 8180,
                        "PublishedPort": 8180
                    }
                ]
            },
            "Ports": [
                {
                    "Protocol": "tcp",
                    "TargetPort": 8180,
                    "PublishedPort": 8180
                }
            ],
            "VirtualIPs": [
                {
                    "NetworkID": "aak1haq3dk71xyzj07kej1kni",
                    "Addr": "10.255.0.6/16"
                }
            ]
        }
    }
]
```

**Comment by aaronlehmann** [comment]
ping @mrjana

**Comment by jfbibeau** [comment]
@aaronlehmann I think you meant to tag this as area/swarm ?

**Comment by aaronlehmann** [comment]
@jfbibeau: Thanks, corrected it.

**Comment by mavenugo** [comment]
The container is connected to the network specific bridge with hairpin=off by default. With ipvs based load-balancing, the VIP is translated back into the container IP and the forwarding is prevented due to  hairpin=off.

A possible fix for this problem is to set the service VIP as the secondary IP of the container. This will also solve the `ping service-name` issue.

**Comment by cpuguy83** [comment]
Closed by #24573
